# HW1 Solution, DATSCI W261, Master Solution

FirstName LastName  
Email address  
Time of Initial Submission: 9:21 PM EST, Monday, January 18, 2016  
Time of **Resubmission**: 8:38 AM EST, Friday, January 22, 2016 
W261-3, Spring 2016  
Week 1 Homework

**Acknowledgements**: Jake Ryland Williams, Liang Dai, Nick Hamlin, James G. Shanahan, Patrick Ng, amongst others 

### |Useful References:
- **[Original Assignment Instructions](https://www.dropbox.com/sh/jylzkmauxkostck/AAA2pH0cTvb0zDrbbbze3zf-a/hw1_instructions.txt?dl=0)**
- [Wikipedia explaination of Naive Bayes document classification](https://en.wikipedia.org/wiki/Naive_Bayes_classifier#Document_classification)
- [Original paper describing the background of the Enron email corpus](http://www.aueb.gr/users/ion/docs/ceas2006_paper.pdf)
- [Documentation for Scikit-Learn implementation of Naive Bayes](http://scikit-learn.org/stable/modules/naive_bayes.html)
- [Stanford NLP Group's explaination of Naive Bayes algorithm](http://nlp.stanford.edu/IR-book/html/htmledition/properties-of-naive-bayes-1.html)

#/Users/jshanahan/anaconda/bin/ipython notebook&

### HW1.0.0: Define big data. Provide an example of a big data problem in your domain of expertise.
Data too large to fit on or be processed by a single machine or tradional techniques is big.
Big data is also well characterized by the four V's: Volume, Variety, Velocity, and Veracity.
In short there is no one definition or definiing characteristic of big data,
which is exemplified by the range of domains from which one can draw examples. 
Other possible talking points for this question: PROCESSING:  Think of your laptop that gets overwhelmed with 3-4 gig of data (disk space is 1TB) 
STORAGE:  Laptop : 1 TB
THROUGH-PUT 1TB would take 3 hours to read it using your laptop. Is this satisfactory? NO!

For the quantitative linguist, the text generated by social media is a prime example of big data---it 
is free form in nature (variety) and generated in real time (velocity) by millions of individuals simultaneously
(volume) whose signals are very often muddied by automatons and spammers (veracity).

### HW1.0.1: In 500 words (English or pseudo code or a combination) describe how to estimate the bias, the variance, the error for a test dataset T when using polynomial regression models of degree 1, 2,3, 4,5 are considered. How would you select a model?
Estimating bias and variance for polynomial models on a dataset with unknown underlying functional form
will generally be predicated on a bootstrapping procedure, creating an ensemble of models for each polynomial degree.
The model variance will be calculated as the average deviation from the average model (for each degree).
Measuring bias is unfortunately somewhat less direct, as one cannot separate the irreducible error 
(noise) present in the data. To determine an estimate of model bias, one can compute the average 
squared error (across an ensemble) from the testing data and subtract off the model variance, 
rendering a value that is the sum of irreducible error and squared bias. Ultimately, to choose a model degree
one can minimize the overall error, but should also be sure to keep a balance between bias and variance indivdiually,
aiming to keep both as low as possible.


###HW1.0.1.
*In 500 words (English or pseudo code or a combination) describe how to estimate the bias, the variance, the irreducible error for a test dataset T when using polynomial regression models of degree 1, 2,3, 4,5 are considered. How would you select a model?*

In the case of regression problems where we estimate mean squared error (MSE), we can decompose the overall expected error in terms of the squared bias, variance, and irreducible error:

$$
E[(\hat{y}-y)^2]=(E[\hat{y}]-y_{true})^2+E[\hat{y}-E[\hat{y}]]^2+E[(y_{true}-y)^2]
$$

$$E[(y-h(x))2]  = Bias[h(x)]^2 + Variance[h(x)] + +E[(y_{true}-y)^2]
$$

Where the h(x) is the model built from each boostrap sample of the test/heldout dataset (see details below).

 
The first term on the RHS is the squared bias, which measures the average error of the model.  The second term is the variance, which measures how much our model's predictions vary from one training set to another.  The final term represent the irreducible error; the variation between our model and reality that we are unable to do anything about (thus the name "irreducible error").

We'd like to choose the model that minimizes both bias and variance. However, we don't know the true function from which T was derived, so calculating this directly doesn't work. In practice, we can calculate the  Expected Error (and its correspending squared bias, and variance) through boostrap sampling of the available training data  (say N bootsrap samples). Using the boostrap training samples, for each hyperparamter (polynomial regression models of different degrees from 1 to M) setting for the training algorithm in question,  we  create N regression models for each degree of polynomial. Each model is then  used to make predictions over the test data. The resulting predictions can then be used  to calculate a biased-squared term, a variance term and the Expected Mean Squared error for the test set (for each candidate hyperparamter setting). The model that has the lowest combined variance AND  squared bias will be the one we choose as best model. In the event that multiple models produce the minimum combination of bias and variance, we should choose the simplest (in this case, the model using the lowest degree polynomial).

Generally, as we increase the degree of the polynomial we use, our bias will drop because we'll fit the points in our training data more closely.  However, this will likely come at a cost of increased variance, as the higher degree polynomials will also mean that our model will be comes less likely to generalize well to new, unseen data points.  Consequently, our ability to successfully classify our training data will fall as our model complexity rises, but our performance on our test sample will begin to suffer as we overfit to the training data.  Referring to the following graph, it is clear that the combine sum of Bias squared and Variance E[MSE] is lowest for a polynomial regression model. This happens to coincide with the E[MSE] with respect to the test set. Note the test error in this graph includes the irreducible error here since we know the true function (in this artifical example). For more details on this example see https://theclevermachine.wordpress.com/2013/04/21/model-selection-underfitting-overfitting-and-the-bias-variance-tradeoff/.


![title](bias-variance-tradeoff.png)

We can use the following pseudocode for these calculations (other variations are also possible such as having an independent test dataset):

In [ ]:
# CASE 1 in a simulated world where we know the ground truth
#HW1.0.1 Pseudocode
# Given 
# A training data samples of the form (X, yObserved) that is generated from a noisy function fNoisy
# and a test set has the form X, y_true in our artificial world that is generated from 
# Running our simulation generates a new column of predictions for each row of the form
#     X, y_true, h_star1, ...h_star_n,
# 
for model in models:
    #this is the bagging step needed to calculate variance
    #where n is some constant (like 50)
    for iteration from 1:n 
        generate a sample of data from our noisy function 
        Train model using train_data
        h_star=predict results for test_data
    h_bar=calculate average prediction across all iterations
    #y_true is the vector of true classes in the test_data
    bias=h_bar-y_true 
    variance=sum((h_bar-h_star)^2)/n #in practice, one would need to go through each iteration to compute this
    noise=mean((y_true-h_star)^2) #As with variance, this needs to be calculated across all iterations
    
choose model that minimizes (bias^2+variance)  

In [ ]:
#CASE 2 in the real world when we do NOT know the true target function

#HW1.0.1 Pseudocode
# Given training data of the form (X, yObserved) 
# y_true is true value but in most real world problems we do NOT know the ground truth 
# So in practice we set y_true =  yObserved and set noise =0
for model in models:
    #this is the bagging step needed to calculate variance
    #where n is some constant (like 50)
    for iteration from 1:n 
        Split training data randomly into train_data and test_data
        Train model using train_data
        h_star=predict results for test_data
    h_bar=calculate average prediction across all iterations
    #y_true is the vector of true classes in the test_data
    bias=h_bar-y_true 
    variance=sum((h_bar-h_star)^2)/n #in practice, one would need to go through each iteration to compute this
    #noise=mean((y_true-h_star)^2) #As with variance, this needs to be calculated across all iterations
    noise=0 #set the noise =0 since we can NOT estimate it
choose model that minimizes (bias^2+variance)  

### HW1.1: Read through the provided control script (pNaiveBayes.sh) and all of its comments. When you are comfortable with their purpose and function, respond to the remaining homework questions below.
No response required.

In [6]:
## HW 1.1 Code

#Display contents of pNaiveBayes.sh (it's convenient to keep everything in one notebook)
!cat pNaiveBayes.sh
!echo ""
!echo "Question 1.1: DONE"

## pNaiveBayes.sh
## Author: Jake Ryland Williams
## Usage: pNaiveBayes.sh m wordlist
## Input:
##       m = number of processes (maps), e.g., 4
##       wordlist = a space-separated list of words in quotes, e.g., "the and of"
##
## Instructions: Read this script and its comments closely.
##               Do your best to understand the purpose of each command,
##               and focus on how arguments are supplied to mapper.py/reducer.py,
##               as this will determine how the python scripts take input.
##               When you are comfortable with the unix code below,
##               answer the questions on the LMS for HW1 about the starter code.

## collect user input
m=$1 ## the number of parallel processes (maps) to run
wordlist=$2 ## if set to "*", then all words are used

## a test set data of 100 messages
data="enronemail_1h.txt" 

## the full set of data (33746 messages)
# data="enronemail.txt" 

## 'wc' determines the number of lines in the data
## 'perl -pe' rege

##HW1.2: Provide a mapper/reducer pair that, when executed by pNaiveBayes.sh will determine the number of occur- rences of a single, user-specified word. Examine the word “assistance” and report your results.
Note that we use the same mapper for each of parts 1.2-1.5 if we structure the code
so that records are printed out at the email ID level, i.e., the email IDs should be the keys
(this will help with classification later). In fact, this part (1.2) is the only one for which we will
need to construct a different reducer (since output is aggregated word counts and not a classification).

In [17]:
%%writefile mapper.py
#!/usr/bin/python
## mapper.py
## Author: Jake Ryland Williams
## Description: mapper code for HW1.2-1.5

import sys
import re
count = 0

## collect user input
filename = sys.argv[1]
findwords = re.split(" ",sys.argv[2].lower())
        
## open the input file
with open (filename, "r") as myfile:
    ## loop over the lines
    for line in myfile.readlines( ) :

        ## check to see if we are counting up a specified set of words
        ## and if so initialize the list
        ALLWORDS=False
        counts = {}        
        if len(findwords) == 1 and findwords[0] == "*":
            ALLWORDS=True
        else:
            for word in findwords:
                counts[word] = 0
                counts[word] = 0
        
        ## strip off line breaks
        line = re.sub('\n','',line)
        ## make sure the format of the line is correct
        if re.match("(.*?)\t(.*?)\t(.*?)\t(.*?)",line):
            ## store the line's data
            [ID,SPAM,SUBJECT,CONTENT] = re.split("\t",line)            
            DAT = CONTENT.lower()
            counts["SPAM"] = SPAM
            if DAT != "NA":
                ## gather words by a regex split
                words = re.split('[^A-Za-z\-\']+',DAT)
                ## loop over words
                for word in words:
                    ## clean up each word
                    neword = re.sub('^\'+','',word)
                    neword = re.sub('^\-+','',neword)
                    neword = re.sub('\-+$','',neword)                    
                    neword = re.sub('\'+','\'',neword)
                    neword = re.sub('\-+','-',neword)
                    ## if there is a word left after the regex, count it
                    if re.match("[a-zA-Z]",neword):                        
                        if ALLWORDS:
                            if neword in counts.keys():
                                counts[neword] += 1
                            else:
                                counts[neword] = 1
                        elif neword in counts.keys():
                            counts[neword] += 1

                dicStr = str(counts)
                ## print out the counts
                print ID + '\t' + dicStr

Overwriting mapper.py


In [13]:
%%writefile reducer.py
#!/usr/bin/python
## reducer.py
## Author: Jake Ryland Williams
## Description: reducer code for HW1.2

import sys
import re
import math

countfiles = sys.argv[1:len(sys.argv)]

## set up local objects
counts = {}

## loop over count files
for filename in countfiles:
    with open (filename, "r") as myfile:
        ## loop over the lines
        for line in myfile.readlines():
            ## clean up data
            line = re.sub('\n','',line)
            ## ignore email ID for now
            [ID,countStr] = re.split("\t",line)
            ## convert the countStr back to a dictionary of counts from the email
            IDcounts = eval(countStr)
            for word in IDcounts.keys():
                ## ignore SPAM key, which associates to binary email class value
                if word != "SPAM":
                    counts.setdefault(word,0)
                    counts[word] += int(IDcounts[word])
                    
for word in counts.keys():
    print "The word \"" + word + "\" appeared " + str(counts[word]) + " times in the dataset."

Overwriting reducer.py


### Make scripts executable

In [18]:
!chmod +x mapper.py; chmod +x reducer.py

###The call for HW1.2:
Note that the word assistance appears only 9 times total in the email's contents,
and 1 time in the email's subjects. This implementation and those below focus strictly on the content.

In [19]:
!./pNaiveBayes.sh 4 "assistance"; cat enronemail_1h.txt.output

The word "assistance" appeared 9 times in the dataset.


##HW1.3: Provide a mapper/reducer pair that, when executed by pNaiveBayes.sh will classify the email messages by a single, user-specified word. Examine the word “assistance” and report your results. 
Here, our reducer must take the word counts from the individual emails passed along from the mapper,
store them in memory by email ID key (for classification), and aggregate them to produce a likelihood function.

We include 2 versions of HW1.3: 

-- one without smoothing and without  log probabilities; 
-- one with smoothing and log probabilities

In [1]:
%%writefile mapper.py
#!/usr/bin/python
#HW 1.3 - Mapper function
#This mapper function will send one line for every instance of every word to the reducer. This approach, while easier to write and debug, is unlikely to be the best choice for a larger scale implementation because of the large volume of data that would have to be sent to the reducers. A potentially more-streamlined alternative would be to add a "combiner" step at the end of the mapper that would send one line for each word-email combination (E.G. Key:email-word-flag, Value:count).
#In addition, if we only care about generating the conditional probabilities for each word and don't need to classify all the training emails, we could simplify the implementation even more and not send the email contents themselves to the reducer. Not only would this decrease throughput, but it would also dramatically reduce the amount of information that the reducer would need to store in memory. In this situation, our mapper could simply emit words along with their conditional class counts. I have not implemented this in this homework, but we'll want to keep this in mind for the future.
 
#HW 1.3 - Mapper Function Code
import sys
import re
WORD_RE = re.compile(r"[\w']+") #Compile regex to easily parse complete words
filename = sys.argv[1]
findwords = sys.argv[2].lower() 
with open (filename, "r") as myfile:
    for num,line in enumerate(myfile.readlines()):
        fields=line.split('\t') #parse line into separate fields
        subject_and_body=" ".join(fields[-2:]).strip()#parse the subject and body fields from the line, and combine into one string
        words=re.findall(WORD_RE,subject_and_body) #create list of words
        for word in words:
            #This flag indicates to the reducer that a given word should be considered
            #by the reducer when calculating the conditional probabilities
            flag=0
            if word in findwords:
                flag=1 
                
            #This will send one row for every word instance to the reducer.
            print fields[0]+'\t'+fields[1]+'\t'+word+'\t1\t'+str(flag)

Overwriting mapper.py


In [2]:

%%writefile reducer.py
#!/usr/bin/python
#HW 1.3 - Reducer function
#The reducer maintains two associative arrays. 
#The first stores information about each word, including how 
#many times it appears in spam and ham messages, as well as if it's been flagged in the mapper. 
#The  second stores information about emails, including whether it is marked as spam, as well as a list of 
#words it contains. 
#As described above, a more scalable solution that does not need to maintain 
#all the contents of the emails in memory for classification could simply calculate conditional 
#probabilities "lazily" and only store the running probability values rather than the words themselves.
#Once all the data has arrived from the mappers, the array containing words is updated with the calculated 
#conditional probabilities of spam and ham. At this point, the model is "trained". Finally, these 
#conditional probabilities are reapplied to the word lists associated with each email to make the 
#final spam/ham classification.
#Note: I have also included (in the comments) an alternative calculation for conditional probabilities that applies a Laplace smoothing approach, since I'd already implemented it before the assignment instructions were updated. I've done the same thing with the log probability calculations.

#HW 1.3 - Reducer Function Code
from __future__ import division #Python 3-style division syntax is much cleaner
import sys
from math import log
 
words={}
emails={}
spam_email_count=0 #number of emails marked as spam
spam_word_count=0 #number of total (not unique) words in spam emails
ham_word_count=0 #number of total (not unique) words in ham emails
flagged_words=[]
for chunk in sys.argv[1:]:
    with open (chunk, "r") as myfile:
        for i in myfile.readlines():
            
            #parse the incoming line
            result=i.split("\t")
            email=result[0]
            spam=int(result[1])
            word=result[2]
            flag=int(result[4])
            
            #initialize storage for word/email data
            if word not in words.keys():
                words[word]={'ham_count':0,'spam_count':0,'flag':flag}
            if email not in emails.keys():
                emails[email]={'spam':spam,'word_count':0,'words':[]}
                if spam==1:
                    spam_email_count+=1
                
            #store word data 
            if spam==1:
                words[word]['spam_count']+=1
                spam_word_count+=1
            else:
                words[word]['ham_count']+=1
                ham_word_count+=1
            
            if flag==1 and word not in flagged_words:
                flagged_words.append(word)
                
            #store email data 
            emails[email]['words'].append(word)
            emails[email]['word_count']+=1
 
#Calculate stats for entire corpus
prior_spam=spam_email_count/len(emails)
prior_ham=1-prior_spam
vocab_count=len(words)#number of unique words in the total vocabulary
            
for k,word in words.iteritems():
    #These versions calculate conditional probabilities WITH Laplace smoothing.  
    #word['p_spam']=(word['spam_count']+1)/(spam_word_count+vocab_count)
    #word['p_ham']=(word['ham_count']+1)/(ham_word_count+vocab_count)
    
    #Compute conditional probabilities WITHOUT Laplace smoothing
    word['p_spam']=(word['spam_count'])/(spam_word_count)
    word['p_ham']=(word['ham_count'])/(ham_word_count)
 
#At this point the model is now trained, and we can use it to make our predictions
for j,email in emails.iteritems():
    
    #Log versions - previously used, but removed for now
    #p_spam=log(prior_spam)
    #p_ham=log(prior_ham)
    
    p_spam=prior_spam
    p_ham=prior_ham
    for word in email['words']:
        if word in flagged_words:
            try:
                #p_spam+=log(words[word]['p_spam']) #Log version - No longer used
                p_spam*=(words[word]['p_spam'])
            except ValueError:
                continue #This means that words that do not appear in a class will use the class prior
            try:
                #p_ham+=log(words[word]['p_ham']) #Log version - No longer used
                p_ham*=(words[word]['p_ham'])
            except ValueError:
                continue          
    if p_spam>p_ham:
        spam_pred=1
    else:
        spam_pred=0
        
    print j+'\t'+str(email['spam'])+'\t'+str(spam_pred)
 

Overwriting reducer.py


In [22]:
#Run our HW 1.3 code and check the results in the output file
!chmod a+x mapper.py reducer.py
!./pNaiveBayes.sh 5 "assistance"
!echo "HW 1.3 - Results"
!cat enronemail_1h.txt.output

HW 1.3 - Results
0010.2003-12-18.GP	1	0
0010.2001-06-28.SA_and_HP	1	1
0001.2000-01-17.beck	0	1
0018.1999-12-14.kaminski	0	0
0005.1999-12-12.kaminski	0	1
0011.2001-06-29.SA_and_HP	1	1
0008.2004-08-01.BG	1	1
0009.1999-12-14.farmer	0	0
0017.2003-12-18.GP	1	0
0011.2001-06-28.SA_and_HP	1	1
0015.2001-07-05.SA_and_HP	1	1
0015.2001-02-12.kitchen	0	1
0009.2001-06-26.SA_and_HP	1	1
0017.1999-12-14.kaminski	0	0
0012.2000-01-17.beck	0	0
0003.2000-01-17.beck	0	0
0004.2001-06-12.SA_and_HP	1	0
0008.2001-06-12.SA_and_HP	1	0
0007.2001-02-09.kitchen	0	1
0016.2004-08-01.BG	1	0
0015.2000-06-09.lokay	0	0
0005.1999-12-14.farmer	0	1
0016.1999-12-15.farmer	0	0
0013.2004-08-01.BG	1	1
0005.2003-12-18.GP	1	1
0012.2001-02-09.kitchen	0	0
0003.2001-02-08.kitchen	0	1
0009.2001-02-09.kitchen	0	0
0006.2001-02-08.kitchen	0	1
0014.2003-12-19.GP	1	0
0010.1999-12-14.farmer	0	0
0010.2004-08-01.BG	1	0
0014.1999-12-14.kaminski	0	0
0006.1999-12-13.kaminski	0	0
0011.1999-12-14.farmer	0	1
0013.1999-12-14.kaminski	0	1
0001.2001-0

In [1]:
#HW 1.3-1.6 Training Error Function

from __future__ import division

def calculate_training_error(pred, true):
    """Calculates the training error given a vector 
    of predictions and a vector of true classes"""
    
    num_wrong=0
    for i in zip(pred,true):
        if i[0]!=i[1]: #If predicted value doesn't equal true value, increment our count
            num_wrong+=1
            
    #Divide number of incorrect examples by total number of examples in the data
    print "Training error: "+str(num_wrong/len(pred))
    
#HW 1.3 Evaluation Code
# To install pandas: at the SHELL prompt type conda install pandas 
import pandas as pd #Use pandas to quickly read results from our output file
#conda install pandas 
def eval_1_3():
    with open('enronemail_1h.txt.output','rb') as f:
        mr_data=pd.read_csv(f, sep='\t', header=None)
    print "Multinomial NB Results via Poor-Man's MapReduce Implementation using 'Assistance' only" 
    calculate_training_error(mr_data[1],mr_data[2])

eval_1_3()

Multinomial NB Results via Poor-Man's MapReduce Implementation using 'Assistance' only
Training error: 0.38


### HW1.3: Provide a mapper/reducer pair that, when executed by pNaiveBayes.sh will classify the email messages by a single, user-specified word. Examine the word “assistance” and report your results. 

**VERSION 2: WITH smoothing and log probabilities** 

Here, our reducer must take the word counts from the individual emails passed along from the mapper,
store them in memory by email ID key (for classification), and aggregate them to produce a likelihood function.
This will totally be overkill for this part (1.3), since the single-word naive Bayes classifier will simplify
to a classification by the prior, and classify all emails to be the class that there was 
more of in the training set (HAM). However for brevity, and since the procedure is generalizable, 
we will use the more complicated design below here as well as for HW1.4 and HW1.5.

In [5]:
%%writefile reducer.py
#!/usr/bin/python
## reducer.py
## Author: Jake Ryland Williams
## Description: reducer code for HW1.3-1.5

import sys
import re
import math

countfiles = sys.argv[1:len(sys.argv)]

## set up local objects
counts = {}
counts["HAM"] = {}
counts["SPAM"] = {}
counts["SPAM"]["T-RECORDS"] = 0 ## for the prior
counts["HAM"]["T-RECORDS"] = 0 ## for the prior
counts["SPAM"]["T-WORDS"] = 0 ## for the class conditional prob. denom.
counts["HAM"]["T-WORDS"] = 0 ## for the class conditional prob. denom.
IDS = []

## loop over count files
for filename in countfiles:
    with open (filename, "r") as myfile:
        ## loop over the lines
        for line in myfile.readlines():
            ## clean up data
            line = re.sub('\n','',line)

            [ID,countStr] = re.split("\t",line)
            IDcounts = eval(countStr)
            SPAM = int(IDcounts["SPAM"])

            ## do some work on the prior
            if SPAM:
                classKey = "SPAM"
                nclassKey = "HAM"
            else:
                classKey = "HAM"
                nclassKey = "SPAM"
            counts[classKey]["T-RECORDS"] += 1

            ## store the record's word counts for classification later
            counts[ID] = IDcounts
            IDS.append(ID)

            ## build up the class-level word counts for the likelihood function
            for word in IDcounts.keys():
                if word != "SPAM":

                    ## store the count
                    count = int(IDcounts[word])
                    
                    ## initialize (with smoothing) in the classes, if necessary
                    if word not in counts[classKey].keys():
                        counts[classKey][word] = 1 ## +1 for Laplace smoothing
                        counts[nclassKey][word] = 1 ## +1 for Laplace smoothing
                        counts[classKey]["T-WORDS"] += 1 
                        counts[nclassKey]["T-WORDS"] += 1

                    ## add to the word's class counts
                    counts[classKey][word] += count
                    counts[classKey]["T-WORDS"] += count

## compute the prior
Ps = float(counts["SPAM"]["T-RECORDS"])/(float(counts["SPAM"]["T-RECORDS"]) + float(counts["HAM"]["T-RECORDS"])) 
Ph = 1-Ps

## initialize components of confusion matrix
TP = 0
FP = 0
TN = 0
FN = 0

## loop over the records-level data, applying the bayesian filter to each
for ID in IDS:

    SPAM = int(counts[ID]["SPAM"])
    lP_ratio = math.log(Ps) - math.log(Ph)

    for word in counts[ID].keys():
        if word != "SPAM":
            ## compute the class conditional probabilities
            sPro = float(counts["SPAM"][word])/float(counts["SPAM"]["T-WORDS"])
            hPro = float(counts["HAM"][word])/float(counts["SPAM"]["T-WORDS"]) 

            ## build up the log likelihood ratio
            lP_ratio += counts[ID][word]*math.log(sPro)
            lP_ratio -= counts[ID][word]*math.log(hPro)

    ## compute the posterior distribution
    try:
        P_ratio = math.exp(lP_ratio)
        Psd = (P_ratio) / (P_ratio + 1)                    
    except OverflowError:
        Psd = 1
    Phd = 1 - Psd
    if (Psd > Phd):
        guess = 1
    else:
        guess = 0
    if guess == SPAM:
        if SPAM:
            TP += 1
        else:
            TN += 1
    else:
        if SPAM:
            FN += 1
        else:
            FP += 1

    ## print out the resulting classification
    print ID+"\t"+str(SPAM)+"\t"+str(guess)+"\t"+str(Psd)+"\t"+str(Phd)

## report confusion matrix statistics

TPR = float(TP) / (float(TP) + float(FN))
FPR = float(FP) / (float(FP) + float(TN))
ACC = (float(TP) + float(TN)) / (float(TP) + float(TN) + float(FP) + float(FN))
print ""
print "========== Summary =========="
print "TPR = " + str(TPR)
print "FPR = " + str(FPR)
print "ACC = " + str(ACC)
print ""

Overwriting reducer.py


### Make scripts executable

In [6]:
!chmod +x mapper.py; chmod +x reducer.py

### The call for HW1.3:

In [7]:
!./pNaiveBayes.sh 4 "assistance"; cat enronemail_1h.txt.output | tail -5

========== Summary ==========
TPR = 0.136363636364
FPR = 0.0357142857143
ACC = 0.6



Doing the learning and classification steps in on Map-Reduce job is acceptable for HW1.

Strictly speaking, going forward, we would have a **second map-reduce job** for doing classification. For classification we would strive to load the entire Naive Bayes classifier model into memory of the mapper. The mapper would calculate the Pr(Class|Example) for each class and do the argmax based classification and output the predicted class, actual class (for now; later we could use combiners). The reducer would then gather classification records and compute metrics such as misclassification rate and output the resulting metric values. Note we should use only a single reducer in the Classification Map-Reduce Job here.


## HW1.4: Provide a mapper/reducer pair that, when executed by pNaiveBayes.sh will classify the email messages by a list of one or more user-specified words. Examine the words “assistance”, “valium”, and “enlargementWithATypo” and report your results.
In this part, no new code is necessary, since mapper.py and reducer.py have been written in the generalized fashion.
2 flavors of solution are given here: 
- one without smoothing and log probabilites; 
- one with smoothing and log probabilities

In [2]:

%%writefile mapper.py
#!/usr/bin/python
#HW 1.4 - Mapper function
#This mapper function works very similarly to the implementation in 1.3. The only difference is that it enables iteration through a list of words (provided as arguments) for flagging for inclusion in the conditional probability calculation.

 
#HW 1.4 - Mapper Function
import sys
import re
WORD_RE = re.compile(r"[\w']+")
filename = sys.argv[1]
findwords = sys.argv[2].lower().split() 
with open (filename, "r") as myfile:
    for num,line in enumerate(myfile.readlines()):
        fields=line.split('\t') #parse line into separate fields
        subject_and_body=" ".join(fields[-2:]).strip()#parse the subject and body fields from the line, and combine into one string
        words=re.findall(WORD_RE,subject_and_body)
        for word in words:
            flag=0
            if word in findwords:
                flag=1
            print fields[0]+'\t'+fields[1]+'\t'+word+'\t1\t'+str(flag)

Overwriting mapper.py


In [3]:
%%writefile reducer.py
#!/usr/bin/python
#HW 1.4 - Reducer function
#This reducer is almost exactly the same as in Problem 1.3. The only difference is not in the code itself, but in the fact that it receives more than one flagged word from the mapper. Because the flagged words are tracked via a list, the reducer doesn't care how many flagged words it receives. It will incorporate all of them into the conditional probability calculation.
#Note: Again, the code for Laplace smoothing and log probability is included as comments, but is not used in the final implementation.

 
#HW 1.4 - Reducer Function
from __future__ import division
import sys
from math import log
 
#
emails={} #Associative array to hold email data
words={} #Associative array for word data
 
spam_email_count=0 #number of emails marked as spam
spam_word_count=0 #number of total (not unique) words in spam emails
ham_word_count=0 #number of total (not unique) words in ham emails
flagged_words=[] #list of flagged words to include in conditional probability calculation
for chunk in sys.argv[1:]:
    with open (chunk, "r") as myfile:
        for i in myfile.readlines():
            
            #parse the line
            result=i.split("\t")
            email=result[0]
            spam=int(result[1])
            word=result[2]
            flag=int(result[4])
            
            #initialize storage for word/email data
            if word not in words.keys():
                words[word]={'ham_count':0,'spam_count':0,'flag':flag}
            if email not in emails.keys():
                emails[email]={'spam':spam,'word_count':0,'words':[]}
                if spam==1:
                    spam_email_count+=1
                
            #store word data 
            if spam==1:
                words[word]['spam_count']+=1
                spam_word_count+=1
            else:
                words[word]['ham_count']+=1
                ham_word_count+=1
            
            if flag==1 and word not in flagged_words:
                flagged_words.append(word)
                
            #store email data 
            emails[email]['words'].append(word)
            emails[email]['word_count']+=1
 
#Calculate stats for entire corpus
prior_spam=spam_email_count/len(emails)
prior_ham=1-prior_spam
vocab_count=len(words)#number of unique words in the total vocabulary
            
for k,word in words.iteritems():
    #These versions calculate conditional probabilities WITH Laplace smoothing.  
    #word['p_spam']=(word['spam_count']+1)/(spam_word_count+vocab_count)
    #word['p_ham']=(word['ham_count']+1)/(ham_word_count+vocab_count)
    
    #Compute conditional probabilities WITHOUT Laplace smoothing
    word['p_spam']=(word['spam_count'])/(spam_word_count)
    word['p_ham']=(word['ham_count'])/(ham_word_count)
 
#At this point the model is now trained, and we can use it to make our predictions
for j,email in emails.iteritems():
    
    #Log versions - no longer used
    #p_spam=log(prior_spam)
    #p_ham=log(prior_ham)
    
    p_spam=prior_spam
    p_ham=prior_ham
    
    for word in email['words']:
        if word in flagged_words:
            try:
                #p_spam+=log(words[word]['p_spam']) #Log version - no longer used
                p_spam*=words[word]['p_spam']
            except ValueError:
                pass #This means that words that do not appear in a class will use the class prior
            try:
                #p_ham+=log(words[word]['p_ham']) #Log version - no longer used
                p_ham*=words[word]['p_ham']
            except ValueError:
                pass
    if p_spam>p_ham:
        spam_pred=1
    else:
        spam_pred=0
        
    print j+'\t'+str(email['spam'])+'\t'+str(spam_pred)
 

Overwriting reducer.py


In [4]:
#Run our HW 1.4 code and check the results in the output file
!chmod a+x mapper.py reducer.py
!./pNaiveBayes.sh 5 "assistance valium enlargementWithATypo"
!echo "HW 1.4 - Results"
!cat enronemail_1h.txt.output

HW 1.4 - Results
0010.2003-12-18.GP	1	0
0010.2001-06-28.SA_and_HP	1	1
0001.2000-01-17.beck	0	0
0018.1999-12-14.kaminski	0	0
0005.1999-12-12.kaminski	0	1
0011.2001-06-29.SA_and_HP	1	0
0008.2004-08-01.BG	1	0
0009.1999-12-14.farmer	0	0
0017.2003-12-18.GP	1	0
0011.2001-06-28.SA_and_HP	1	1
0015.2001-07-05.SA_and_HP	1	0
0015.2001-02-12.kitchen	0	0
0009.2001-06-26.SA_and_HP	1	0
0017.1999-12-14.kaminski	0	0
0012.2000-01-17.beck	0	0
0003.2000-01-17.beck	0	0
0004.2001-06-12.SA_and_HP	1	0
0008.2001-06-12.SA_and_HP	1	0
0007.2001-02-09.kitchen	0	0
0016.2004-08-01.BG	1	0
0015.2000-06-09.lokay	0	0
0005.1999-12-14.farmer	0	0
0016.1999-12-15.farmer	0	0
0013.2004-08-01.BG	1	1
0005.2003-12-18.GP	1	0
0012.2001-02-09.kitchen	0	0
0003.2001-02-08.kitchen	0	0
0009.2001-02-09.kitchen	0	0
0006.2001-02-08.kitchen	0	0
0014.2003-12-19.GP	1	0
0010.1999-12-14.farmer	0	0
0010.2004-08-01.BG	1	0
0014.1999-12-14.kaminski	0	0
0006.1999-12-13.kaminski	0	0
0011.1999-12-14.farmer	0	0
0013.1999-12-14.kaminski	0	0
0001.2001-0

In [5]:
#HW 1.4 - Evaluation code
def eval_1_4():
    with open('enronemail_1h.txt.output','rb') as f:
        mr_data=pd.read_csv(f, sep='\t', header=None)
    print "Multinomial NB Results via Poor-Man's MapReduce Implementation using 'Assistance valium EnlargementWithATypo'" 
    calculate_training_error(mr_data[1],mr_data[2])

eval_1_4()

Multinomial NB Results via Poor-Man's MapReduce Implementation using 'Assistance valium EnlargementWithATypo'
Training error: 0.37


Unsurprisingly, the addition of a few more words improves performance slightly, but not enough to make this an effective model for real spam classification.

## HW1.4: Provide a mapper/reducer pair that, when executed by pNaiveBayes.sh will classify the email messages by a list of one or more user-specified words. Examine the words “assistance”, “valium”, and “enlargementWithATypo” and report your results.

**VERSION 2: WITH smoothing and log probabilities** 

In this part, no new code is necessary, since mapper.py and reducer.py have been written in the generalized fashion.
We will continue to use the mapper from 1.2, and the reducer from 1.3.

### The call for HW1.4:

In [8]:
!./pNaiveBayes.sh 4 "assistance valium enlargementWithATypo"; cat enronemail_1h.txt.output | tail -5

========== Summary ==========
TPR = 0.159090909091
FPR = 0.0357142857143
ACC = 0.61



##HW1.5: Provide a mapper/reducer pair that, when executed by pNaiveBayes.sh will classify the email messages by all words present.
In this part, no new code is necessary, since mapper.py and reducer.py have been written in the generalized fashion.
We will continue to use the mapper from 1.2, and the reducer from 1.3.

### The call for HW1.5:

In [9]:
!./pNaiveBayes.sh 4 "*"; cat enronemail_1h.txt.output | tail -5

========== Summary ==========
TPR = 1.0
FPR = 0.178571428571
ACC = 0.9



## HW1.6: Benchmark your code with the Python SciKit-Learn implementation of Naive Bayes.

In [37]:
import re
import numpy as np
from sklearn.naive_bayes import MultinomialNB, BernoulliNB
from sklearn.feature_extraction.text import CountVectorizer
    
emails = []
classes = []
    
with open ("enronemail_1h.txt", "r") as myfile:
    for line in myfile.readlines():
        
        [ID,SPAM,SUBJECT,CONTENT] = re.split("\t",line)            
        DAT = CONTENT.lower()            
        emails.append(DAT)
        classes.append(SPAM)
    
classes = np.array(classes)
    
# Create features for train and dev data 
vectorizer = CountVectorizer(min_df=3)
trainingData = vectorizer.fit_transform(emails)
    
print "\nTraining Error"    
classifier = BernoulliNB()
classifier.fit(trainingData, classes) 
print "  SkLearn Bernoulli NB:\t\t", 1-classifier.score(trainingData,classes)
    
classifier = MultinomialNB()
classifier.fit(trainingData, classes) 
print "  SkLearn Multinomial NB:\t", 1-classifier.score(trainingData,classes)
   
print "  HW1.5 Multinomial NB:\t\t", 1-0.9


Training Error
  SkLearn Bernoulli NB:		0.16
  SkLearn Multinomial NB:	0.03
  HW1.5 Multinomial NB:		0.1


#### HW 1.6 - Summary of Results

| Model                                                                      | Training Error |
|----------------------------------------------------------------------------|----------------|
| Multinomial NB, Scikit-Learn Implementation                                | 0.03            |
| Bernoulli NB, Scikit-Learn Implementation                                  | 0.16           |
| Multinomial NB, MapReduce implementation                                   | 0.1           |


Note that the observed error varies between the different implementations of multinomial naive Bayes.
Overall, both version perform similarly, though differ at the level of feature extraction.
Since SkLearn's default feature extractor relys only on the '\w' regular expression, 
it includes digits in the definitions of words, while ignoring hyphenations and contractions.
This generally results in a larger, more specific vocabulary, given the larger range of acceptable characters
(and the introduction of junk pieces of words left by splitting at apostraphes and hyphens),
that can overfit and result in the reported lower error, since we are testing on the training set.

Note as well that between the two SkLearn implementations of Naive Bayes, the Multinomial calssifier 
outperforms the Bernoulli classifier. This is likely due to the fact that the Bernoulli classifier
ignores frequency of occurrence, making it difficult to compare documents of disparate sizes 
(which the emails are very easily seen to be).